### PR TITLE
Implement osu!'s beat snapping grid

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/6473

Same as the PR - implementation + tests. Not integrated yet - integration is not far off but there's a few further prerequisites/cleanups to do.

### CODE REVIEW

File: osu.Game.Rulesets.Osu.Tests/TestSceneOsuBeatSnapGrid.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        [TestCase(1.5f, 150)]
+        public void TestSliderMultiplier(float multiplier, float expectedSpacing)
+        {
+            AddStep($"set speed multiplier = {multiplier}", () => editorBeatmap.BeatmapInfo.BaseDifficulty.SliderMultiplier = multiplier);
+            createGrid();
+            AddAssert($"spacing = {expectedSpacing}", () => Precision.AlmostEquals(expectedSpacing, grid.DistanceSpacing));
+        }
+
+        [Test]
+        public void TestCursorInCentre()
```
Comment (bdach) [inline osu.Game.Rulesets.Osu.Tests/TestSceneOsuBeatSnapGrid.cs]: This test is giving me a consistent crash locally in the visual test runner on the last step (same stack trace as the CI build is reporting for it), which is weird, as all of the others run fine to the end.

**Review by bdach — COMMENTED** [review]
I like how this is panning out, works really nice. I have some suggestions/notes in comments, and some general notes that are probably outside of the scope of this - you might be aware of them or not, but I thought I'd want to mention them:

- I feel that for the finer spacings (like on `TestBeatDivisor(16)`) the usefulness of the grid is greatly diminished by the density of the circular rulers. I'd consider some sort of hover behaviour for it to be easier to tell visually which circle is being snapped to.
- The rulers seem to be aliased badly, which is quite jarring - see magnified image below:
  image

**Comment by bdach** [comment]
A random thought I had on the problem of denser grids not really helping much for snapping was applying a notion of [hysteresis](https://en.wikipedia.org/wiki/Hysteresis#User_interface_design) to the grid lines. I'm not sure if I'll be able to explain clearly how this applies, but here goes.

The rulers could be designed such that the snapping range of the currently active ruler is not exactly half of the distance between rulers, but a little more than that (so like 75%). To switch to snap to another ruler you'd have to get closer than halfway (following the number above, within 25% of the distance to it), at which point it becomes the active ruler and has the wider margin from that point on. Could even combine that with a hover effect that that makes the active ruler glow and show its snapping range for clear UX.

This is a suggestion that I haven't tried at all and most likely out of scope of this, I guess we'd have to try it and see if it feels ok.

**Comment by smoogipoo** [comment]
Hysteresis is a little bit out of scope for now, just want to get things working and usable for now.

**Comment by smoogipoo** [comment]
> The rulers seem to be aliased badly, which is quite jarring

Can't really address this one easily, but it would be nice to fix. It's a framework-side fix to a component that doesn't really support AA in the form we usually do.

**Review by bdach — DISMISSED** [review]
Looks ready to go as far as I can tell :shipit:

**Review by peppy — APPROVED** [review]

**Pull request merged**